src/vs/workbench/contrib/terminal/electron-browser/terminalInstance.ts

```
@@ -739,14 +739,33 @@ export class TerminalInstance implements ITerminalInstance {
 		}
 		if (this._processManager) {
 			this._processManager.dispose(immediate);
+		} else {
+			// In cases where there is no associated process (for example executing an extension callback task)
+			// consumers still expect on onExit event to be fired. An example of this is terminating the extension callback
+			// task. There is no exit code at this point, so firing undefined is appropriate.
+			this._onExit.fire(undefined);
 		}
+
 		if (!this._isDisposed) {
 			this._isDisposed = true;
 			this._onDisposed.fire(this);
 		}
 		this._disposables = lifecycle.dispose(this._disposables);
 	}
 
+	public finishedWithRenderer(result?: number): void {
+		// The use of this API is for cases where there is no backing process
+		// behind a terminal instance (such as when executing an custom execution task).
+		// There is no associated string, error text, etc, as the consumer of the renderer
+		// can simply output  the text to the renderer themselves.
+		// All this code does is handle the "wait on exit" condition.
+		if (!this.shellLaunchConfig.isRendererOnly) {
+			return;
```

Let's throw an Error, this should never happen